catroid/src/main/java/org/catrobat/catroid/transfers/project/ProjectDownloadService.kt

```diff
@@ -0,0 +1,182 @@
+
+        const val UPDATE_PROGRESS_EXTRA = "progress"
+
+        const val UPDATE_PROGRESS_CODE = 1
+        const val ERROR_CODE = 2
+        const val SUCCESS_CODE = 3
+    }
+
+    override fun onHandleIntent(intent: Intent?) {
+        val downloadIntent = intent
+            ?: return logWarning("Called ProjectDownloadService with null intent - aborting")
+        val projectName = downloadIntent.getStringExtra(EXTRA_DOWNLOAD_NAME)
+            ?: return logWarning("Called ProjectDownloadService with null projectName -  aborting")
+        val url = downloadIntent.getStringExtra(EXTRA_URL)
+            ?: return logWarning("Called ProjectDownloadService without url - aborting")
+        val resultReceiver = downloadIntent.getParcelableExtra<ResultReceiver>(EXTRA_RESULT_RECEIVER)
+                ?: return logWarning("Called ProjectDownloadService without url - aborting")
+
+        val zipFileString = File(File(CACHE_DIR, TMP_DIR_NAME), DOWNLOAD_FILE_NAME).absolutePath
+        val destinationFile = File(zipFileString)
+
+        if ((destinationFile.parentFile.isDirectory or destinationFile.parentFile.mkdirs()).not()) {
+            ToastUtil.showError(this, R.string.error_project_download)
+            return
+        }
+
+        val id = downloadIntent.getIntExtra(EXTRA_NOTIFICATION_ID, 0)
+        val statusBarNotificationManager = StatusBarNotificationManager.getInstance()
+        val notification = statusBarNotificationManager.getNotification(this, id)
+
+        // startForeground binds the notification to the service
+        // since stopForeground(STOP_FOREGROUND_DETACH) works only after version O and we don't
+        // need startForeground before, just exclude it from previous versions
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            startForeground(id, notification)
+        }
+
+        val context = this
+        CatrobatServerCalls(CatrobatWebClient.client)
+            .downloadProject(
+                url,
+                destinationFile,
+                object : DownloadSuccessCallback {
+                    override fun onSuccess() {
+                        downloadSuccessCallback(
+                            context,
+                            projectName,
+                            destinationFile,
+                            resultReceiver,
+                            id
+                        )
+                    }
+                },
+                object : DownloadErrorCallback {
+                    override fun onError(code: Int, message: String) {
+                        downloadErrorCallback(context, resultReceiver, id)
+                    }
+                },
+                object : DownloadProgressCallback {
+                    override fun onProgress(progress: Long) {
+                        downloadProgressCallback(context, resultReceiver, id, progress)
+                    }
+                }
+            )
```

Comment:
you don't need to store context in a separate var. Use `this@ProjectDownloadService` instead.